src/ring_theory/unique_factorization_domain.lean

```diff
@@ -1,18 +1,150 @@
+  principal ideals in an integral domain.
+   -/
+class DCC_dvd (α : Type*) [comm_monoid_with_zero α] : Prop :=
+(well_founded_dvd_not_unit : well_founded (λ a b : α, a ≠ 0 ∧ ∃ x, ¬is_unit x ∧ b = a * x))
+
+export DCC_dvd (well_founded_dvd_not_unit)
+
+instance is_noetherian_ring.DCC_dvd [integral_domain α] [is_noetherian_ring α] :
+  DCC_dvd α :=
+⟨by simp only [ideal.span_singleton_lt_span_singleton.symm];
+   exact inv_image.wf (λ a, ideal.span ({a} : set α)) (well_founded_submodule_gt _ _)⟩
+
+instance polynomial.DCC_dvd [integral_domain α] [DCC_dvd α] : DCC_dvd (polynomial α) :=
+{ well_founded_dvd_not_unit := begin
+    classical,
+    apply rel_hom.well_founded, swap,
+    exact (prod.lex_wf (with_top.well_founded_lt $ with_bot.well_founded_lt nat.lt_wf)
+      _inst_2.well_founded_dvd_not_unit),
+    exact {
+      to_fun := λ p, (if p = 0 then ⊤ else ↑p.degree, p.leading_coeff),
+      map_rel' := λ a b ⟨ane0, ⟨c, ⟨not_unit_c, bac⟩⟩⟩, begin
+        rw bac, rw polynomial.degree_mul, rw if_neg ane0, by_cases c = 0,
+        { simp only [h, if_true, polynomial.leading_coeff_zero, eq_self_iff_true, ne.def, mul_zero],
+          apply prod.lex.left _ _ (lt_of_le_of_ne le_top _), apply with_top.coe_ne_top, },
+        { simp only [h, ane0, if_false, ne.def, polynomial.leading_coeff_mul, or_self, mul_eq_zero],
+          rename h cne0, by_cases c.degree = 0,
+        { simp only [h, add_zero, ne.def, polynomial.leading_coeff_mul],
+          apply prod.lex.right, split, {rw polynomial.leading_coeff_eq_zero, apply ane0},
+          use c.leading_coeff, split, swap, {refl},
+          contrapose! not_unit_c, rw polynomial.is_unit_iff, use c.leading_coeff,
+          split, {exact not_unit_c}, rw polynomial.leading_coeff,
+          convert (polynomial.eq_C_of_degree_eq_zero h).symm,
+          apply polynomial.nat_degree_eq_of_degree_eq_some h,},
+        { apply prod.lex.left, rw polynomial.degree_eq_nat_degree cne0 at *,
+          rw [with_top.coe_lt_coe, polynomial.degree_eq_nat_degree ane0,
+            ← with_bot.coe_add, with_bot.coe_lt_coe],
+          apply lt_add_of_pos_right, apply nat.pos_of_ne_zero, contrapose! h, rw h, refl, }, }
+      end }
+  end }
+end prio
+
+namespace DCC_dvd
+
+variables [comm_monoid_with_zero α]
+open associates nat
+
+theorem of_DCC_dvd_associates
+  [DCC_dvd (associates α)]: DCC_dvd α :=
+⟨begin
+  refine (surjective.well_founded_iff mk_surjective _).2 DCC_dvd.well_founded_dvd_not_unit,
+  intros, rw mk_dvd_not_unit_mk_iff
+end⟩
+
+variables [DCC_dvd α]
+
+@[priority 100]
+instance DCC_dvd_associates : DCC_dvd (associates α) :=
+⟨begin
+  refine (surjective.well_founded_iff mk_surjective _).1 DCC_dvd.well_founded_dvd_not_unit,
+  intros, rw mk_dvd_not_unit_mk_iff
+end⟩
+
+theorem well_founded_associates : well_founded ((<) : associates α → associates α → Prop) :=
+subrelation.wf (λ x y, dvd_not_unit_of_lt) DCC_dvd.well_founded_dvd_not_unit
```

Comment:
Could you add a theorem `iff_well_founded_associates` that uses `dvd_not_unit_iff_lt` to show `well_founded_associates` is an iff in the nice cases?

Reply:
It exists, but the proof is not as nice, and it's not quite in the namespace. I can fix these things.